### PR TITLE
Repository upkeep & automation idea

### PR DESCRIPTION
**Pull request by Zlaticanin**
`make setup-env`
`make source-env`
`make configure`

### CODE REVIEW

File: bootstrap/terraform/redis.tf [hunk 1]
```diff
@@ -0,0 +1,20 @@
+resource "null_resource" "setup_redis_docker" {
+  triggers = {
+    always = timestamp()
+  }
+
+  provisioner "local-exec" {
+    command = "cd ../../scripts && ./gen-test-certs.sh && docker-compose up -d"
+  }
+}
```
Comment (maxcoulombe) [inline bootstrap/terraform/redis.tf:9]: We could add a resource to handle the make env-teardown command so the docker image gets cleaned-up:
```
resource "null_resource" "teardown_redis_docker" {
  triggers = {
    always = timestamp()
  }

  provisioner "local-exec" {
    when = destroy
    command = "docker stop vault_redis_test && docker rm vault_redis_test"
  }
}
```

File: scripts/compose.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      - REDIS_TLS_AUTH_CLIENTS=no
+    ports:
+      - '6379:6379'
+    volumes:
+      - 'redis_data:/bitnami/redis/data'
+      - ./tests/tls:/tls
+
+volumes:
+  redis_data:
+    driver: local
```
Comment (maxcoulombe) [inline scripts/compose.yml:24]: This docker compose file could be replaced with a classic docker run command in a bash file or embeded directly in the Terraform null_resource so users do not need to install docker compose as an extra dependency:
```
docker stop vault_redis_test || true
docker rm vault_redis_test || true

docker run -d \
  -p 6379:6379 \
  -u "${UID}:${GID}" \
  -e ALLOW_EMPTY_PASSWORD=false \
  -e REDIS_USERNAME=default \
  -e REDIS_PASSWORD=the-strong-one \
  -e REDIS_DISABLE_COMMANDS=FLUSHDB,FLUSHALL \
  -e REDIS_TLS_CERT_FILE=/tls/redis.crt \
  -e REDIS_TLS_KEY_FILE=/tls/redis.key \
  -e REDIS_TLS_CA_FILE=/tls/ca.crt \
  -e REDIS_TLS_ENABLED=yes \
  -e REDIS_TLS_PORT=6379 \
  -e REDIS_TLS_AUTH_CLIENTS=no \
  -v redis_data:/bitnami/redis/data \
  -v $(pwd)/tests/tls:/tls \
  --volume-driver local \
  --name vault_redis_test \
  docker.io/bitnami/redis:6.2
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	@sh -c "'$(CURDIR)/scripts/gofmtcheck.sh'"
+
+.PHONY: fmt
+fmt:
+	gofumpt -l -w . && cd bootstrap/terraform && terraform fmt
+
+.PHONY: setup-env
+setup-env:
+	cd bootstrap/terraform && terraform init && terraform apply -auto-approve
+
```
[the remark this answers is not in the document]
Comment (Zlaticanin) [reply Makefile:32]: Good catch 👍

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	CGO_ENABLED=0 go build -o bin/$(PLUGIN_NAME) cmd/$(PLUGIN_NAME)/main.go
+	cp bin/$(PLUGIN_NAME) $(PLUGIN_DIR)/$(PLUGIN_NAME)
+
+.PHONY: test
+test:
+	CGO_ENABLED=0 go test -v ./... $(TESTARGS) -timeout=20m
+
+.PHONY: testacc
+testacc:
+	ACC_TEST_ENABLED=1 CA_CERT_FILE=$(TEST_REDIS_CACERT_PATH) REDIS_HOST=$(TEST_REDIS_HOST) REDIS_TLS=true CGO_ENABLED=0 go test -v ./... $(TESTARGS) -timeout=20m
```
[the remark this answers is not in the document]
Comment (Zlaticanin) [reply Makefile]: I think most of the repos are using `VAULT_ACC`. Also interested to hear from others if we should leave it like this, or change it back for consistency. 
But maybe if we decide to implement this for other plugins, we can update this var as well.

File: README.md [hunk 5]
```diff
@@ -41,41 +56,100 @@ Code: 400. Errors:
...
-```bash
-$ SHA256=$(shasum -a 256 plugins/redis-database-plugin | cut -d' ' -f1)
+### Environment Set Up
 
-$ vault secrets enable database
+To test `go test` will execute a set of basic tests against the `docker.io/redis:latest` Redis database image. To test against different Redis images, for example 5.0-buster, set the `REDIS_VERSION=5.0-buster` environment variable. If you want to run the tests against a local Redis installation or an already running Redis container, set the environment variables `REDIS_HOST` before executing, as well as `REDIS_TLS`, `CA_CERT_FILE` for acceptance tests.
+
+**Note:** The tests assume that the Redis database instance has a default user with the following ACL settings user default on `nopass ~* +@all`. If not you will need to align the Administrator username and password with the pre-set values in the [redis_test.go](https://github.com/hashicorp/vault-plugin-database-redis/blob/main/redis_test.go) file.
+
+Set `VAULT_ACC` to execute all of the tests. A subset of tests can be run using the command `go test -run TestDriver/Init` for example.
```
[the remark this answers is not in the document]
Comment (Zlaticanin) [reply README.md]: 👍

File: README.md [hunk 6]
```diff
@@ -41,41 +56,100 @@ Code: 400. Errors:
...
+The setup script tries to find and use available Redis credentials from the environment. You can configure Redis credentials by using or editing the provider defined `./bootstrap/terraform/redis.tf` with your desired set of credentials.
+
+To set up the test container:
+
+```sh
+$ make setup-env
+...
+Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
+
+$ source ./bootstrap/terraform/local_environment_setup.sh
```
[the remark this answers is not in the document]
Comment (Zlaticanin) [reply README.md]: Good suggestion 👍

File: scripts/compose.yml (hunk 2 above)
Comment (Zlaticanin) [reply scripts/compose.yml]: Good idea. I created a new bash file, so now the docker container is created using a classic run command. I deleted the compose file.

**Pull request closed without merging**